warehouse/oidc/models/semaphore.py

```diff
@@ -0,0 +1,240 @@
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import typing
+
+from typing import Any, Self
+from uuid import UUID
+
+from sqlalchemy import ForeignKey, String, UniqueConstraint, and_, exists
+from sqlalchemy.dialects.postgresql import UUID as PG_UUID
+from sqlalchemy.orm import Mapped, Query, mapped_column
+
+from warehouse.oidc.errors import InvalidPublisherError
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models._core import (
+    CheckClaimCallable,
+    OIDCPublisher,
+    PendingOIDCPublisher,
+    check_claim_binary,
+    check_existing_jti,
+)
+
+if typing.TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+
+SEMAPHORE_OIDC_ISSUER_URL_SUFFIX = ".semaphoreci.com"
+
+
+def _check_sub(
+    ground_truth: str,
+    signed_claim: str,
+    _all_signed_claims: SignedClaims,
+    **_kwargs,
+) -> bool:
+    # Semaphore's sub claim contains:
+    # org:<org-name>:project:<project-uuid>:repo:<owner/repo>:ref_type:<type>:ref:<ref>
+    # We verify that it contains the expected repo
+
+    if not signed_claim:
+        return False
+
+    # Extract the repo portion from the sub claim
+    if ":repo:" not in signed_claim:
+        return False
+
+    parts = signed_claim.split(":repo:")
+    if len(parts) < 2:  # pragma: no cover
+        # This should be unreachable since we already checked ":repo:" is in the string
+        return False
+
+    # The repo value is between :repo: and the next :
+    repo_and_rest = parts[1]
+    repo_parts = repo_and_rest.split(":", 1)
+    repo_in_sub = repo_parts[0]
+
+    if not repo_in_sub:
+        return False
```

Comment:
I think this can be simplified a bit:
```suggestion
    # There should only be one ':repo:' in the sub:
    if signed_claim.count(":repo:") != 1:
        return False

    _, repo_and_rest = signed_claim.split(":repo:", 1)
    repo_in_sub, _ = repo_and_rest.split(":", 1)

    if not repo_in_sub:
        return False
```

Reply:
Simplified